cassandane/Cassandane/Cyrus/ALPN.pm

```diff
@@ -0,0 +1,244 @@
+#  are met:
+#
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in
+#     the documentation and/or other materials provided with the
+#     distribution.
+#
+#  3. The name "Fastmail Pty Ltd" must not be used to
+#     endorse or promote products derived from this software without
+#     prior written permission. For permission or any legal
+#     details, please contact
+#      FastMail Pty Ltd
+#      PO Box 234
+#      Collins St West 8007
+#      Victoria
+#      Australia
+#
+#  4. Redistributions of any form whatsoever must retain the following
+#     acknowledgment:
+#     "This product includes software developed by Fastmail Pty. Ltd."
+#
+#  FASTMAIL PTY LTD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
+#  INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY  AND FITNESS, IN NO
+#  EVENT SHALL OPERA SOFTWARE AUSTRALIA BE LIABLE FOR ANY SPECIAL, INDIRECT
+#  OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
+#  USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+#  TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
+#  OF THIS SOFTWARE.
+#
+
+package Cassandane::Cyrus::ALPN;
+use strict;
+use warnings;
+use Cwd qw(abs_path);
+use Data::Dumper;
+
+use lib '.';
+use base qw(Cassandane::Cyrus::TestCase);
+use Cassandane::Util::Log;
+
+sub new
+{
+    my $class = shift;
+
+    my $config = Cassandane::Config->default()->clone();
+    $config->set(tls_server_cert => '@basedir@/conf/certs/cert.pem',
+                 tls_server_key => '@basedir@/conf/certs/key.pem');
+
+    my $self = $class->SUPER::new({
+        config => $config,
+        install_certificates => 1,
+        services => [ 'imap', 'imaps' ],
+    }, @_);
+
+    $self->needs('dependency', 'openssl');
+    $self->needs('dependency', 'openssl_alpn');
+
+    return $self;
+}
+
+sub set_up
+{
+    my ($self) = @_;
+    $self->SUPER::set_up();
+}
+
+sub tear_down
+{
+    my ($self) = @_;
+    $self->SUPER::tear_down();
+}
+
+sub do_imap_starttls
+{
+    my ($self, $talk, $alpn_map) = @_;
+    my $ca_file = abs_path("data/certs/cacert.pem");
+
+    my $starttls_tag = $talk->{CmdId};
+
+    $talk->_imap_cmd('starttls', 0, 'starttls');
+
+    die $talk->get_last_error()
+        if $talk->get_last_completion_response() ne 'ok';
+
+    IO::Socket::SSL->start_SSL($talk->{Socket},
+                               SSL_ca_file => $ca_file,
+                               SSL_verifycn_scheme => 'none',
+                               SSL_alpn_protocols => $alpn_map,
+    );
+
+    if (ref $talk->{Socket} ne 'IO::Socket::SSL') {
+        # STARTTLS failed! Rummage inside Mail::IMAPTalk to put it back into
+        # a sane state, because the tag having two responses confuses it
+
```

Comment:
After re-reading RFC 2595, I'm wondering if imapd, pop3d, etc should just call shutdown() if STARTTLS fails

Reply:
Hmm, it doesn't really give any clear direction about what to do when the TLS negotiation fails, does it. [This paragraph](https://www.rfc-editor.org/rfc/rfc2595#section-2.5) is about as close as it gets:

> Both the client and server MUST check the result of the STARTTLS command and subsequent TLS negotiation to see whether acceptable authentication or privacy was achieved.  Ignoring this step completely invalidates using TLS for security.  The decision about whether acceptable authentication or privacy was achieved is made locally, is implementation-dependent, and is beyond the scope of this document.

I think that's giving us permission to handle it however we like. I would argue that a client which has issued STARTTLS expects to get a secure connection, and may not correctly handle the case where the TLS negotiation fails and the connection remains unencrypted.  A naive client might just send their password over the wire at this point.  That being the case, the server should drop the connection rather than give them the opportunity to do so.

I'm not sure I'm comfortable bundling such a change into this PR though, though it is highlighting the need for it.  Guess I'll add it to the pile.

Reply:
I can workin that if you like.

Reply:
Sure, I actually just wrote up a linear task for it.  It's in the triage queue if you wanna grab it.

Reply:
@elliefm See #5133

Reply:
@elliefm See #5134

Reply:
Great, thanks! I'll rebase on that once it lands, and fix my tests to match